### PR TITLE
Refactor: Extract serialization into modular architecture and wire into existing pipeline

### PR DESCRIPTION
**Pull request by TooTallNate**
## Summary

Extracts the monolithic 1892-line `serialization.ts` into a modular architecture with clear separation of concerns, and **completes the migration** by wiring the existing `serialization.ts` to use the new modules as its foundation. The event-replay runtime is already exercising the new modular code via the `dehydrate*`/`hydrate*` functions (see [verification](#verification) below).

### What changed

New `packages/core/src/serialization/` directory with:

| Module | Purpose |
|--------|------------|
| `types.ts` | `FormatPrefix`, `SerializableSpecial`, `Reducers`, `Revivers` types |
| `codec.ts` | `Codec` interface — enables future alternative formats (CBOR, JSON) |
| `codec-devalue.ts` | Devalue implementation of the Codec interface |
| `format.ts` | Format prefix encode/decode/peek (`devl`, `encr`) |
| `encryption.ts` | Composable encrypt/decrypt layer (wraps AES-256-GCM) |
| `errors.ts` | Shared `formatSerializationError` helper used by all modes |
| `reducers/common.ts` | Date, Error, Map, Set, URL, BigInt, typed arrays, DOMException, WorkflowFunction, etc. |
| `reducers/class.ts` | Class/Instance with `WORKFLOW_SERIALIZE`/`DESERIALIZE` |
| `reducers/step-function.ts` | StepFunction with closure variables |
| `workflow.ts` | **Workflow mode**: `serialize()` / `deserialize()` — synchronous, no encryption |
| `step.ts` | **Step mode**: `serialize()` / `deserialize()` — async, with encryption |
| `client.ts` | **Client mode**: `serialize()` / `deserialize()` — async, with encryption |
| `index.ts` | Re-exports public API |

### Migration completed

The existing `serialization.ts` has been refactored to **import from the new modules** instead of duplicating code:

- **Format prefix system** (`encodeWithFormatPrefix`, `decodeFormatPrefix`, `peekFormatPrefix`, `isEncrypted`) → imported from `serialization/format.ts`
- **Types** (`SerializableSpecial`, `Reducers`, `Revivers`, `FormatPrefix`) → imported from `serialization/types.ts`
- **Common reducers/revivers** (Date, Error, Map, Set, typed arrays, etc.) → imported from `serialization/reducers/common.ts`
- **Class reducers/revivers** → imported from `serialization/reducers/class.ts`
- **Step function reducers/revivers** → imported from `serialization/reducers/step-function.ts`
- **Encryption helpers** (`maybeEncrypt`/`maybeDecrypt`) → now delegate to `serialization/encryption.ts`
- **`formatSerializationError`** → imported from `serialization/errors.ts`

The 8 `dehydrate*`/`hydrate*` functions now delegate to the modular mode modules (`clientModule.serialize`, `workflowModule.deserialize`, `stepModule.serialize`, `stepModule.deserialize`) with stream + Request/Response reducers/revivers passed as `CodecOptions.extraReducers`/`extraRevivers`.

**File shrunk from 1892 → 1421 lines** (~470 lines of duplicate code removed).

Stream handling (`getSerializeStream`, `getDeserializeStream`, `WorkflowServerReadableStream`, `WorkflowServerWritableStream`) and mode-specific Request/Response/stream reducers/revivers remain in `serialization.ts` since they're tightly coupled to runtime concerns (world, ops arrays, flushable pipes, WritableStream symbols).

### Design decisions

- **Codec interface**: Abstracts the serialization format so CBOR/JSON can be added without changing mode-specific modules
- **Composable encryption**: `encrypt(serialize(value))` instead of encryption tangled into serialize
- **Mode-specific modules**: Each mode (`workflow`, `step`, `client`) composes its own reducer/reviver set from shared building blocks
- **Workflow mode is synchronous**: No async, no encryption — designed to run inside a sandboxed VM (e.g. the forthcoming QuickJS snapshot runtime)
- **CodecOptions**: Threading `global`, `extraReducers`, `extraRevivers` through the Codec interface preserves VM-context `instanceof` checks and lets the legacy `dehydrate*`/`hydrate*` functions inject stream/Request/Response handlers without duplicating the core reducer set

### Verification

The event-replay runtime is **already** using the new modular code. Every runtime call site goes `runtime → legacy dehydrate/hydrate wrapper → modular serialize/deserialize`:

| Runtime file | Function called | Delegates to |
|---|---|---|
| `runtime/start.ts` | `dehydrateWorkflowArguments` | `clientModule.serialize` |
| `runtime/run.ts` | `hydrateWorkflowReturnValue` | `clientModule.deserialize` |
| `runtime/runs.ts` | `hydrateWorkflowArguments` | `workflowModule.deserialize` |
| `runtime/step-handler.ts` | `hydrateStepArguments`, `dehydrateStepReturnValue` | `stepModule.deserialize`, `stepModule.serialize` |
| `runtime/suspension-handler.ts` | `dehydrateStepArguments` | `stepModule.serialize` |
| `runtime/resume-hook.ts` | `hydrateStepArguments`, `dehydrateStepReturnValue` | `stepModule.deserialize`, `stepModule.serialize` |
| `workflow.ts` | `hydrateWorkflowArguments`, `dehydrateWorkflowReturnValue` | `workflowModule.deserialize`, `stepModule.serialize` |
| `workflow/hook.ts` (in-VM) | `hydrateStepReturnValue` | `workflowModule.deserialize` |
| `step.ts` (in-VM) | `hydrateStepReturnValue` | `workflowModule.deserialize` |

The only `v1Compat` fast path for legacy spec-version-1 runs still uses `stringify` + `revive` directly — that's the pre-refactor behavior preserved intentionally for compatibility with runs that predate format-prefixed binary payloads.

### New tests

**138 focused unit tests** covering:
- `types.ts`: `isFormatPrefix` type guard, `SerializationFormat` constants
- `format.ts`: `encodeWithFormatPrefix`, `decodeFormatPrefix`, `peekFormatPrefix`, `isEncrypted` (including edge cases for empty/large payloads, legacy data, invalid prefixes)
- `encryption.ts`: `encrypt`/`decrypt` round-trips, no-key passthrough, error on missing key
- `reducers/common.ts`: All common reducers and revivers with round-trip verification
- `reducers/class.ts`: Class/Instance reducers and revivers, error cases (missing classId, missing WORKFLOW_DESERIALIZE)
- `reducers/step-function.ts`: StepFunction reducer with/without closure vars, reviver with mocked WORKFLOW_USE_STEP
- `codec-devalue.ts`: All modes, primitives, complex types, legacy deserialization, StepFunction error in client mode
- `workflow.ts`: All serializable types, format prefix verification, error handling, legacy data
- `step.ts`: Round-trips, encryption, format prefix, legacy data
- `client.ts`: Round-trips, encryption, error handling
- Cross-mode compatibility: All 6 mode combinations (workflow↔step↔client), including encrypted cross-mode
- Edge cases: circular references, deep nesting, undefined values, empty collections, BigInt boundaries

All **729 tests pass** (591 existing + 138 new).

### CODE REVIEW

File: packages/core/src/serialization/client.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  global: Record<string, any> = globalThis
+): Partial<Reducers> {
+  return {
+    ...getCommonReducers(global),
+    ...getClassReducers(),
+    // Note: Stream reducers for client mode need additional parameters
+    // (ops, runId, cryptoKey). These are composed at call sites that
+    // need stream support.
+  };
+}
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/client.ts]: Fixed in 63b21f11 — class/instance reducers now spread before common reducers in all three modes.

File: packages/core/src/serialization/workflow.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+// ---- Reducer/Reviver composition ----
+
+function getWorkflowReducers(
+  global: Record<string, any> = globalThis
+): Partial<Reducers> {
+  return {
+    ...getCommonReducers(global),
+    ...getClassReducers(),
+    ...getStepFunctionReducer(),
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/workflow.ts]: Fixed in 63b21f11 — class/instance reducers now spread before common reducers in all three modes.

File: packages/core/src/serialization/step.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  global: Record<string, any> = globalThis
+): Partial<Reducers> {
+  return {
+    ...getCommonReducers(global),
+    ...getClassReducers(),
+    // Note: Stream reducers for step mode need additional parameters
+    // (ops, runId, cryptoKey). These are composed at call sites that
+    // need stream support. For basic step serialization, common + class
+    // reducers are sufficient.
+  };
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/step.ts]: Fixed in 63b21f11 — class/instance reducers now spread before common reducers in all three modes.

File: packages/core/src/serialization/workflow.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+import { WorkflowRuntimeError } from '@workflow/errors';
+import { DevalueError } from 'devalue';
+import { runtimeLogger } from '../logger.js';
+import { devalueCodec } from './codec-devalue.js';
+import { decodeFormatPrefix, encodeWithFormatPrefix } from './format.js';
+import { getClassReducers, getClassRevivers } from './reducers/class.js';
+import { getCommonReducers, getCommonRevivers } from './reducers/common.js';
+import {
+  getStepFunctionReducer,
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/workflow.ts]: Fixed in 63b21f11 — class/instance reducers now spread before common reducers in all three modes.

File: packages/core/src/serialization/encryption.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * @param key - Encryption key (undefined to skip decryption)
+ * @returns The decrypted inner payload, or the original data if not encrypted
+ */
+export async function decrypt(
+  data: Uint8Array | unknown,
+  key: CryptoKey | undefined
+): Promise<Uint8Array | unknown> {
+  if (!key || !(data instanceof Uint8Array)) return data;
+  if (peekFormatPrefix(data) !== SerializationFormat.ENCRYPTED) return data;
+
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/encryption.ts]: Fixed in 63b21f11 — decrypt() now throws immediately when encrypted data is encountered without a key.

File: packages/core/src/serialization/reducers/common.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        headers: value.headers,
+        body: value.body,
+        duplex: value.duplex,
+      };
+      const responseWritable = value[WEBHOOK_RESPONSE_WRITABLE];
+      if (responseWritable) {
+        data.responseWritable = responseWritable;
+      }
+      return data;
+    },
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/core/src/serialization/reducers/common.ts]: Fixed in 63b21f11 — removed Request/Response from common reducers. They belong in mode-specific modules that can provide matching revivers.

File: packages/workflow/src/internal/serialization.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ * This module re-exports the workflow-mode serialize/deserialize from
+ * @workflow/core. It is designed to be imported by the compiled workflow
+ * bundle (via the SWC plugin or VM bootstrap code) and executed inside
+ * the sandboxed VM environment.
+ *
+ * The serialize/deserialize functions are synchronous and do not use
+ * encryption — encryption is handled on the host side outside the VM.
+ */
+export { serialize, deserialize } from '@workflow/core/serialization/workflow';
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/workflow/src/internal/serialization.ts]: Updated the doc comment in 63b21f11 to explicitly document the Node.js dependencies (node:util, Buffer) and note that polyfills are needed for the QuickJS VM — tracked on the snapshot-runtime branch.

File: packages/core/src/serialization/encryption.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  key: CryptoKey | undefined
+): Promise<Uint8Array | unknown> {
+  // Non-binary data is returned as-is.
+  if (!(data instanceof Uint8Array)) return data;
+
+  const format = peekFormatPrefix(data);
+
+  // If the data is encrypted but no key was provided, fail fast.
+  if (format === SerializationFormat.ENCRYPTED && !key) {
+    throw new Error(
```
Comment (VaguelySerious) [inline packages/core/src/serialization/encryption.ts]: ### AI Review: Blocking

The old `maybeDecrypt` threw `WorkflowRuntimeError` for this case. The new `decrypt()` throws a plain `Error`. Since `maybeDecrypt` now delegates to this function, callers that relied on `WorkflowRuntimeError` semantics get a different error type.

Confirmed via validation test: `maybeDecrypt(encryptedData, undefined)` now throws `Error` instead of `WorkflowRuntimeError`.

Suggestion: Use `WorkflowRuntimeError` here to preserve the original error contract, or document this as a deliberate breaking change.

File: packages/core/src/serialization/format.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    throw new Error(
+      `Data too short to contain format prefix: expected at least ${FORMAT_PREFIX_LENGTH} bytes, got ${data.length}`
+    );
+  }
+
+  const prefixBytes = data.subarray(0, FORMAT_PREFIX_LENGTH);
+  const str = formatDecoder.decode(prefixBytes);
+
+  if (!isFormatPrefix(str)) {
+    throw new Error(
```
Comment (VaguelySerious) [inline packages/core/src/serialization/format.ts:117]: ### AI Review: Blocking

Same error type change as the encryption module: the old `decodeFormatPrefix` in `serialization.ts` threw `WorkflowRuntimeError` with "Unknown serialization format". This now throws a plain `Error` with "Invalid format prefix".

Additionally, the validation semantics changed:
- **Old**: Rejects any prefix not in `Object.values(SerializationFormat)` (only `devl`, `encr`)
- **New**: Accepts any `[a-z0-9]{4}` as valid

This means data with a `cbor` prefix (not yet implemented) now passes through `decodeFormatPrefix` and hits "Unsupported serialization format" at the mode level. Arguably better for forward compatibility, but it is a behavioral change — worth calling out explicitly.

File: packages/core/src/serialization/workflow.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  const { format, payload } = decodeFormatPrefix(data);
+
+  if (format === SerializationFormat.DEVALUE_V1) {
+    return devalueCodec.deserialize(payload, 'workflow');
+  }
+
+  throw new Error(`Unsupported serialization format: ${format}`);
+}
+
+function formatSerializationError(context: string, error: unknown): string {
```
Comment (VaguelySerious) [inline packages/core/src/serialization/workflow.ts]: ### AI Review: Note

`formatSerializationError` is duplicated identically in 4 files: `serialization.ts`, `workflow.ts`, `step.ts`, and `client.ts`. Consider extracting to a shared utility (e.g., `serialization/errors.ts`).

File: packages/core/src/serialization/codec-devalue.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+// ---- Reducer/Reviver composition per mode ----
+
+function getReducersForMode(mode: SerializationMode): Partial<Reducers> {
+  switch (mode) {
+    case 'workflow':
+      // Class/Instance MUST come before common (first-match-wins for Error subclasses)
+      return {
+        ...getClassReducers(),
+        ...getStepFunctionReducer(),
+        ...getCommonReducers(),
```
Comment (VaguelySerious) [inline packages/core/src/serialization/codec-devalue.ts]: ### AI Review: Note

`getCommonReducers()` and `getCommonRevivers()` are called without a `global` parameter, defaulting to `globalThis`. The original code in `serialization.ts` supports VM contexts by passing `global` through to these functions.

This means the modular modules (workflow.ts, step.ts, client.ts) can't be used for cross-VM serialization where constructors differ. The old dehydrate/hydrate path still works correctly, and inside the VM `globalThis` IS the VM's global, but this limitation should be documented — especially since `packages/workflow/src/internal/serialization.ts` re-exports this path for the VM bundle.

File: packages/core/src/runtime/run.ts [hunk 12]
```diff
@@ -217,12 +217,8 @@ export class Run<TResult> {
...
     // the first async transform() call of the deserialize stream.
     const encryptionKey = this.getEncryptionKey();
-    const stream = getExternalRevivers(
-      global,
-      ops,
-      this.runId,
-      encryptionKey
-    ).ReadableStream({
+    const stream = getExternalRevivers(global, ops, this.runId, encryptionKey)
+      .ReadableStream!({
```
Comment (VaguelySerious) [inline packages/core/src/runtime/run.ts:270]: ### AI Review: Nit

The `!` non-null assertion is needed because `getExternalRevivers` now returns `Partial<Revivers>`. While safe at runtime (the function always includes `ReadableStream`), it reduces type safety. Consider a narrower type for the return of `getExternalRevivers` that guarantees the keys it always includes.

File: packages/core/src/serialization/codec.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
+/**
```
Comment (VaguelySerious) [inline packages/core/src/serialization/codec.ts:1]: ### AI Review: Nit

With only one codec implementation (devalue) and no immediate plans for CBOR/JSON, this interface adds a layer of indirection. It's clean and well-scoped, but could be deferred until a second codec is actually needed.

**Review by VaguelySerious — COMMENTED** [review]
AI review: blocking issues found

File: packages/core/src/serialization/encryption.ts (hunk 8 above)
Comment (TooTallNate) [reply packages/core/src/serialization/encryption.ts]: Fixed in e1c35120 — `decrypt()` now throws `WorkflowRuntimeError` instead of plain `Error` to preserve the error contract from the legacy `maybeDecrypt()` implementation.

File: packages/core/src/serialization/format.ts (hunk 9 above)
Comment (TooTallNate) [reply packages/core/src/serialization/format.ts:117]: Good catch on the behavioral change. This is intentional — the open-ended `[a-z0-9]{4}` validation is designed for forward compatibility so new codecs (e.g. `cbor`) can be added without modifying this module. Added a doc comment in e1c35120 explicitly documenting this design choice and the caller's responsibility to check format support.

File: packages/core/src/serialization/workflow.ts (hunk 10 above)
Comment (TooTallNate) [reply packages/core/src/serialization/workflow.ts]: Agreed — extracted into `serialization/errors.ts` in e1c35120. All four copies (serialization.ts, workflow.ts, step.ts, client.ts) now import from the shared utility.

File: packages/core/src/serialization/codec-devalue.ts (hunk 11 above)
Comment (TooTallNate) [reply packages/core/src/serialization/codec-devalue.ts]: Good point. Added a doc comment block in e1c35120 documenting this limitation. Inside the VM `globalThis` is the VM's global so it works correctly there, and the legacy dehydrate/hydrate path in `serialization.ts` still supports passing a custom `global` for full cross-VM compatibility. Threading `global` through the Codec interface is deferred until the snapshot runtime work requires it.

File: packages/core/src/runtime/run.ts (hunk 12 above)
Comment (TooTallNate) [reply packages/core/src/runtime/run.ts:270]: Agreed this is a tradeoff. The `!` assertion is safe at runtime since `getExternalRevivers` always includes `ReadableStream`, but the type doesn't guarantee it. A narrower return type would be cleaner but would require a larger refactoring of the reviver composition system. Leaving as-is for now — the `Partial<Revivers>` return type was a pragmatic choice to avoid fighting the type system during the modular composition.

File: packages/core/src/serialization/codec-devalue.ts (hunk 11 above)
Comment (TooTallNate) [reply packages/core/src/serialization/codec-devalue.ts]: Correction to my earlier reply: after tracing through the full runtime, the `globalThis` default is actually a non-issue for the current runtime. All serialization in the current event replay flow goes through the `dehydrate*/hydrate*` functions in `serialization.ts`, which pass `global` (either `vmGlobalThis` or host `globalThis`) through to the reducer/reviver factories. The modular mode modules (`workflow.ts`, `step.ts`, `client.ts`) are not used in the current runtime at all — they are infrastructure for the future snapshot runtime where serialization would run *inside* the VM sandbox (where `globalThis` IS the VM's global). Updated the comment in codec-devalue.ts to reflect this accurately.

File: packages/core/src/serialization/codec.ts (hunk 13 above)
Comment (TooTallNate) [reply packages/core/src/serialization/codec.ts:1]: Fair point. The Codec interface is forward-looking — the snapshot runtime branch needs to swap in a CBOR codec for the QuickJS VM. Since it's a small interface (3 methods) and doesn't add meaningful complexity, keeping it as scaffolding for that upcoming work.

**Review by VaguelySerious — APPROVED** [review]
